src/PartialModelSelection.cpp

```diff
@@ -1,202 +1,199 @@
+   startingModel.modelSizeAfter = 1;
+   PenaltyModelPair startingPair = PenaltyModelPair(0.0, startingModel); 
+   try{
+      auto insertResult = penaltyModelMap.insert(startingPair);
+      validateInsert(insertResult);
+      prevInsertedPair = insertResult.first; //Set the iterator to the previous insert to the validated insert as we did not error.
+      
+   }
+
+   catch(std::logic_error errorMessage) {
+      std::cout << "Insert failed, validation returned an error!\n";
+   }
+   
+   
+   
+   insertedModels = 0; //This starts at 0 as we exclude the beginning placeholder.
+}
+
+
+void ModelSelectionMap::insert(double newPenalty, Model newModel){
+   //Insert into ourpenaltyModelPair map in the ModelSelectionMap if the newPenalty is not within it.
+   PenaltyModelPair newPair = PenaltyModelPair(newPenalty, newModel);
+   
+   try{
+      auto insertResult = penaltyModelMap.insert(newPair);
+      validateInsert(insertResult);
+      //After inserting, update our previous entry as well. 
+      updatePreviousEntry();
+      prevInsertedPair = insertResult.first; //Set the iterator to the previous insert to the validated insert as we did not error.
+   }
+
+   catch(std::logic_error errorMessage) {
+      std::cout << "Insert failed, validation returned an error!\n";
+   }
+
+   
+   
+   insertedModels++;     
+}
+
+
+void ModelSelectionMap::insert(Model currentModel)
+{
+   //Runs solver for penaltyQuery to insert
+   //std::cout << "Adding model without penaltyQuery \n";
+
+   //Insert pair with createdpenaltyModelPair from solver.
+}
+
+
+
+MinimizeResult ModelSelectionMap::minimize(double penaltyQuery){
+   //Default result if we do not find a matching penaltyQuery.
+   MinimizeResult queryResult = MinimizeResult(); 
+   auto indexPair = penaltyModelMap.find(penaltyQuery);
+   double indexPenalty = indexPair->first;
+   Model indexModel = indexPair->second;
+
+   //If we have no inserted model/penaltyQuery pairs, return the default result from 0 to inf.
+   if(!hasModelsInserted()){
+    std::cout << "Map is empty, returning default min result\n";
+    return queryResult;
+   }
+
+    //If we have models inserted, we need to find a penaltyQuery result that is closest to the queried penaltyQuery
+    indexPair = penaltyModelMap.lower_bound(penaltyQuery);
+
+    //If we found an inserted pair that lies on the queried penalty itself
+    if(indexPenalty == penaltyQuery)
+      //Make a query result to return using the second element of a testedPair, Model. Get it's modelSize. 
+      queryResult = MinimizeResult(std::make_pair(indexModel.modelSize,indexModel.modelSize));
+   
+    //Otherwise, make a range query that does not lie on an inserted pair. 
+    else{
+       queryResult = MinimizeResult(std::make_pair(indexModel.modelSize, indexModel.modelSizeAfter)); //Returning default for now. 
+       queryResult.certain = false;
+    }
+    
+    return queryResult;
+}
+
+
+
+
+double ModelSelectionMap::getNewPenaltyList()
```

Comment:
please delete this method, we don't want to allocate memory for a list, because we will be returning a vector of new penalties in R. so instead please define a C++ function that lets us iterate over new penalties, and then we can use that to create the vector to return in R.